### ISSUE TITLE
Cast to Timestamp does not support ISO Date Time Format

### ISSUE THREAD

**Issue by rupal-s** [trinodb/trino#17313]
The most common ISO Date Time Format yyyy-MM-dd'T'HH:mm:ss.SSSXXX — for example, "2000-10-31T01:30:00.000-05:00". 
However, Trino does not support this: 
`trino> select  cast('2000-10-31T01:30:00.000-05:00' as timestamp);
Query 20230502_070845_00033_xpsfg failed: Value cannot be cast to timestamp: 2000-10-31T01:30:00.000-05:00
`
However, we do support the above value without the 'T'
 
`trino> select  cast('2000-10-31 01:30:00.000-05:00' as timestamp);
          _col0          
-------------------------
 2000-10-31 01:30:00.000 
(1 row)
`

The ISO date time format should be supported as well.
Related PR: https://github.com/trinodb/trino/pull/14166

**Comment by Praveen2112** [comment]
Instead can we use `from_iso8601_timestamp` to convert ISO date time format to timestamp.

**Comment by Praveen2112** [comment]
cc: @martint Is there is any specific reason on not supporting ISO date time format in `CAST` function ?

**Comment by martint** [comment]
> cc: @martint Is there is any specific reason on not supporting ISO date time format in CAST function ?

Indeed. The reason is that CAST is a special construct defined by the SQL specification, and it does not contemplate that format.

**Comment by rupal-s** [comment]
@martint I didn't quite understand. Do we not want to support cast from iso format?

**Comment by martint** [comment]
Not via CAST, as that is not a behavior covered by the SQL specification. We have a set of dedicated functions for that purpose, as @Praveen2112 mentioned above:

* https://trino.io/docs/current/functions/datetime.html?#from_iso8601_date
* https://trino.io/docs/current/functions/datetime.html?#from_iso8601_timestamp
* https://trino.io/docs/current/functions/datetime.html?#from_iso8601_timestamp_nanos
* https://trino.io/docs/current/functions/datetime.html?#to_iso8601

**Comment by rupal-s** [comment]
@martint due to this bug: "Cast to timestamp ignores trailing content it does not recognize" (https://github.com/trinodb/trino/issues/14164), our customers who were used to sending timestamps to be cast in the ISO format like "2000-10-31T01:30:00.000-05:00" thought it was passing (although the code was simply ignoring everything after the 'T')
After the bug fix by the PR https://github.com/trinodb/trino/pull/14166, the above timestamp now throws the error "Value cannot be cast to timestamp:" 
So, to the customer it feels like we have **removed** support for ISO timestamp. 
Given this context, can we consider supporting ISO timestamp for the cast function as well? I would be happy to raise the PR.

**Comment by Praveen2112** [comment]
> can we consider supporting ISO timestamp for the cast function as well

I don't think we can - as the SQL specification doesn't support ISO format. 

cc: @martint Can we have the specification for timestamp cast for future reference ?
